sandbox/samples/cpp/alembicassembly/alembicassembly.cpp

```diff
@@ -0,0 +1,1525 @@
+
+        // multiply the each matrix of the stack from bottom to top
+        for (const asf::Matrix4d& mtx : mtx_stack)
+        {
+            out_mtx = mtx * out_mtx;
+        }
+
+        return out_mtx;
+    }
+
+    // roll into given xform sequence stack and return the flatten one
+    asr::TransformSequence flatten_xform_seq(const std::vector<asr::TransformSequence>& xform_seq_stack)
+    {
+        assert(xform_seq_stack.size() > 0);
+
+        if (xform_seq_stack.size() == 1)
+        {
+            // one item, we don't need to flatten anything
+            return xform_seq_stack.back();
+        }
+        else
+        {
+            // xform sequence we will return
+            asr::TransformSequence out_xform_seq;
+
+            for (const asr::TransformSequence& xform_seq : xform_seq_stack)
+            {
+                out_xform_seq = xform_seq * out_xform_seq;
+            }
+
+            out_xform_seq.optimize();
+            out_xform_seq.prepare();
+
+            return out_xform_seq;
+        }
+    }
+
+    // return if given time samples are linearly time spaced between start_time
+    // and end_time.
+    const bool are_linearly_sampled(const AbcChrono_t start_time,
+                                    const AbcChrono_t end_time,
+                                    const std::vector<AbcChrono_t>& samples)
+    {
+        assert(samples.size() > 1);
+
+        // compute time supposed to be between each samples if samples are
+        // linearly sampled.
+        const AbcChrono_t increm = (end_time - start_time) / (samples.size()-1);
+
+        // start sample
+        AbcChrono_t t_accum = samples[0];
+
+        // get each sample time and compare with a linearly sampled time
+        for (const AbcChrono_t t : samples)
+        {
+            if (asf::feq(t, t_accum))
+            {
+                return false;
+            }
+
+            t_accum += increm;
+        }
+        return true;
+    }
+
+
+    // return sample indices (with border samples) in given shutter
+    //                           open          close
+    // shutter times       ---------|--------------|------------->
+    //                  (floor)  ,<-'              '-->,  (ceil)
+    // returned samples    ------|-----------|---------|--------->
+    std::vector<AbcIndex_t> schema_to_sample_times(
+        const float shutter_open,
+        const float shutter_close,
+        const size_t num_samples,
+        const Abc::TimeSamplingPtr time_sampling)
+    {
+        assert(shutter_open <= shutter_close);
+
+        // the index array we will return
+        std::vector<AbcIndex_t> sample_indices;
+
+        if (num_samples < 2)
+        {
+            // no sample, put the main sample (sample 0)
+            sample_indices.push_back(0);
+        }
+        else  // more than 1 sample
+        {
+            // get floor and ceil sample indices to don't miss any
+            const std::pair<AbcIndex_t, AbcChrono_t>& open_pair =
```

Comment:
Just checking: are you sure that `getFloorIndex()` returns a _reference_ and not a value?

Reply:
No it doesn't. I was doing this to save a copy. Should I remove the ref identifier?

Reply:
> Should I remove the ref identifier?

Yes, because as the code stands right now, you're keeping a reference to a temporary object that no longer exists outside `getFloorIndex()`, i.e. it's broken (even if it may occasionally work).